Add `ViT` example (#3063)
adap/flower: examples/vit-finetune/README.md

# Federated finetuning of a ViT

This example shows how to use Flower's Simulation Engine to federate the finetuning of a Vision Transformer ([ViT-Base-16](https://pytorch.org/vision/main/models/generated/torchvision.models.vit_b_16.html#torchvision.models.vit_b_16)) that has been pretrained on ImageNet. To keep things simple we'll be finetuning it to CIFAR-100, creating 20 partitions using [Flower-datasets](https://flower.ai/docs/datasets/). We'll be finetuning just the exit `head` of the ViT, this means that the training is not that costly and each client requires just ~1.5GB of VRAM (for a batch size of 64 images).

## Running the example

If you haven't cloned the Flower repository already you might want to clone code example and discard the rest. We prepared a single-line command that you can copy into your shell which will checkout the example for you:

```shell
git clone --depth=1 https://github.com/adap/flower.git && mv flower/examples/vit-finetune . && rm -rf flower && cd vit-finetune
```

> This will create a new directory called `simulation-pytorch` containing the following files:

Reviewer: ```suggestion
This will create a new directory called `vit-finetune` containing the following files:
```

```
-- README.md         <- Your're reading this right now
-- main.py           <- Main file that launches the simulation
-- client.py         <- Contains Flower client code and ClientApp
-- server.py         <- Contains Flower server code and ServerApp
-- model.py          <- Defines model and train/eval functions
-- dataset.py        <- Downloads, partitions and processes dataset

> -- pyproject.toml    <- Example dependencies
> -- requirements.txt  <- Example dependencies

Reviewer: ```suggestion
-- pyproject.toml    <- Example dependencies, installable using Poetry
-- requirements.txt  <- Example dependencies, installable using pip
```

+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A     30741      C   python                                     1966MiB |
|    0   N/A  N/A     31419      C   ray::DefaultActor.run                      1536MiB |
|    0   N/A  N/A     31420      C   ray::DefaultActor.run                      1536MiB |
|    0   N/A  N/A     31421      C   ray::DefaultActor.run                      1536MiB |
|    0   N/A  N/A     31422      C   ray::DefaultActor.run                      1536MiB |
+---------------------------------------------------------------------------------------+
```

### Run with Flower Next (preview)

```bash

> flower-simulation --client-app=client:app --server-app=server:app --num-supernodes=20 \
>     --backend-config='{"client_resources": {"num_cpus":4, "num_gpus":0.25}}'

Reviewer: ```suggestion
flower-simulation \
    --client-app=client:app \
    --server-app=server:app \
    --num-supernodes=20 \
    --backend-config='{"client_resources": {"num_cpus":4, "num_gpus":0.25}}'
```
